### PR TITLE
Var dropout

### PR DESCRIPTION
**Pull request by RomainDeleat-besson**

### CODE REVIEW

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+* It is also introducing the *Local Reparameterization trick*
+* Finally, both previous points help to developp *variational dropout*
+
+
+&nbsp;
+
+
+# Introduction
+
+Has it is well known nowadays, deep neural networks can scale to millions of parameters.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:35]: Spelling:
```suggestion
As is well known nowadays, deep neural networks can scale to millions of parameters. 
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+* Finally, both previous points help to developp *variational dropout*
+
+
+&nbsp;
+
+
+# Introduction
+
+Has it is well known nowadays, deep neural networks can scale to millions of parameters. 
+However, this flexbility may leads to overfitting therefore not being robust for new data.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:36]: Conjugation + phrasing:
```suggestion
However, this flexbility may lead to overfitting, and therefore decrease robustness on new data.
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+Various regularization techniques are used in practice such as dropout for instance.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:37]: Phrasing:
```suggestion
Various regularization techniques are used in practice, such as dropout.
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+
+In a previous paper that they cite, it was shown that binary dropout has a Gaussian approximation called Gaussian dropout.
+Based on that, they show that a relationship exists between dropout and Bayesian inference to greatly improve the efficiency of variational Bayesian inference on the **model parameters**. 
+The problem they underline is that :
+> Bayesian methods for inferring a posterior distribution over neural network weights have not yet been shown to outperform simpler methods such as dropout.
+
+
+That's why they proposed a new trick called the *Local reparameterization trick* that will improve the stochastic gradientbased variational inference with minibatches of data.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:46]: Missing space:
```suggestion
That's why they proposed a new trick called the *Local reparameterization trick* that will improve the stochastic gradient based variational inference with minibatches of data. 
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+
+In a previous paper that they cite, it was shown that binary dropout has a Gaussian approximation called Gaussian dropout.
+Based on that, they show that a relationship exists between dropout and Bayesian inference to greatly improve the efficiency of variational Bayesian inference on the **model parameters**. 
+The problem they underline is that :
+> Bayesian methods for inferring a posterior distribution over neural network weights have not yet been shown to outperform simpler methods such as dropout.
+
+
+That's why they proposed a new trick called the *Local reparameterization trick* that will improve the stochastic gradientbased variational inference with minibatches of data. 
+Leading to an optimization speed on the same level as fast dropout.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:47]: This sentence should be rephrased (e.g. by making it the end of the last sentence), because it's not a proper sentence right now.

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+To do so they propose a new estimator where the aim is to have $$Cov \left[ L_i, L_j \right] = 0$$ so that the variance of the stochastic gradients scales as $$1/M$$.
+
+Exemple taken in the paper :
+> Consider a standard fully connected neural network containing a hidden layer consisting of 1000 neurons. 
+This layer receives an $$M$$ x $$1000$$ input feature matrix $$\textbf{A}$$ from the layer below, which is multiplied by a $$1000$$ x $$1000$$ weight matrix $$\textbf{W}$$, before a nonlinearity is applied, i.e. $$ \textbf{B = AW}$$. 
+We then specify the posterior approximation on the weights to be a fully factorized Gaussian, i.e. $$q_\phi(w_{i,j}) =N (\mu_{i,j}, \sigma_{i,j}^2) \forall w_{i,j} \in \textbf{W}$$, which means the weights are sampled as 
+$$w_{i,j} = \mu_{i,j} + \sigma_{i,j} \epsilon_{i,j}$$, with $$\epsilon_{i,j} \sim  \mathcal{N}(0, 1)$$.
+In this case we could make sure that $$Cov [L_i,L_j]=0 $$ by sampling a separate weight matrix $$\textbf{W}$$.
+
+This approach is not computationally efficient. To make it so they applied another trick based on
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:104]: Phrasing:
```suggestion
This approach is not computationally efficient. To make it so they applied another trick:
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+
+Dropout consists to add multiplicative noise to the imput of each layer. For a fully connected layer we have :
+
+$$ \textbf{B} = (\textbf{A} \circ \xi)\theta $$ with $$\xi \sim p(\xi_{i,j})$$
+
+- $$\textbf{A}$$ is the $$M$$ x $$K$$ matrix of input features
+- $$\theta$$ is a $$K$$ x $$L$$ weight matrix
+- $$\textbf{B}$$ is the $$M$$ x $$L$$ output matrix
+
+It was shown on previous paper that using a continuous distribution such as a Gaussian $$\mathcal{N}(1,\alpha)$$ with $$\alpha = p(1-p)$$
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:138]: Fix phrasing + alpha definition:
```suggestion
It was shown in a previous paper that using a continuous distribution such as a Gaussian $$\mathcal{N}(1,\alpha)$$ with $$\alpha = \frac{p}{(1-p)}$$ 
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+
+The local reparameterization trick leads to an estimator that has lower variance and is more computationally efficient.
+
+&nbsp;
+
+
+## Variational Dropout
+
+Dropout consists to add multiplicative noise to the imput of each layer. For a fully connected layer we have :
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:130]: Conjugation + typo:
```suggestion
Dropout consists of adding multiplicative noise to the input of each layer. For a fully connected layer we have :
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+with 
+
+$$\gamma_{m,j} = \sum_{i=1}^{1000} a_{m,i}\mu_{i,j}$$ and $$\delta_{m,j} = \sum_{i=1}^{1000} a_{m,i}^2\sigma_{i,j}^2$$
+
+
+>Rather than sampling the Gaussian weights and then computing the resulting activations, 
+we may thus sample the activations from their implied Gaussian distribution directly, 
+using $$b_{m,j} = \gamma_{m,j} + \sqrt{\delta_{m,j}}\zeta_{m,j}$$, with $$\zeta_{m,j} \sim  \mathcal{N}(0, 1)$$. 
+
+Here, $$\zeta$$ is an $$M$$ x $$1000$$ matrix, so only $$M$$ thousand random variables are needed to be sample instead of $$M$$ million.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:120]: I would rephrase the definition of the complexity like this:
```suggestion
Here, $$\zeta$$ is an $$M \times 1000$$ matrix, so only $$M \times 1000$$ random variables have to be sampled instead of $$M \times 1\,000\,000$$.
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+
+> If the elements of the noise matrix $$\xi$$ are drawn independently from a Gaussian $$N (1, \alpha)$$, the marginal distributions of the activations $$b_{m,j} \in \textbf{B}$$ are Gaussian as well:
+
+$$q_\phi(b_{m,j} \vert \textbf{A}) = N (\gamma_{m,j}, \delta_{m,j}) $$ 
+with 
+$$\gamma_{m,j} = \sum_{i=1}^{K} a_{m,i}\theta_{i,j}$$ and $$\delta_{m,j} = \alpha \sum_{i=1}^{K} a_{m,i}^2\theta_{i,j}^2$$
+
+With this equation, the activations are directly drawn from their (approximate or exact) marginal distributions.
+
+This Gaussian dropout noise can be see as a Bayesian treatment of a neural network with $$\textbf{B = AW}$$ where the **posterior distribution of the weights** is given by a factorized Gaussian with :
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:154]: Conjugation:
```suggestion
This Gaussian dropout noise can be seen as a Bayesian treatment of a neural network with $$\textbf{B = AW}$$ where the **posterior distribution of the weights** is given by a factorized Gaussian with :
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+
+$$q_\phi(w_{i,j}) = \mathcal{N}(\theta_{i,j}, \alpha\theta_{i,j}^2)$$
+
+
+Another section is about variational dropout with correlated weight noise but not developped in this post.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:160]: Clarify that the additional section on variational dropout is in the paper:
```suggestion
The paper includes another section on variational dropout with correlated weight noise, but we do not developp it in this post.
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+
+$$- D_{KL} \left[ q_\phi(w_i) \vert \vert p(w_i) \right] \approx constant + 0.5log(\alpha) + c_1\alpha + c_2\alpha^2 + c_3\alpha^3$$
+
+with
+
+$$ c_1 = 1.16145124, c_2 = -1.50204118 , c_3 = 0.58629921$$
+
+Now that there is a derived dropout's variational objective, 
+they show that maximizing the variational lower bound with respect to $$\alpha$$ will make the hyperparameters $$\alpha$$ and $$\theta$$ adaptative.
+Furthermore, it will be possible to learn a separate dropout rate per layer, per neuron, of even per separate weight.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:197]: Typo:
```suggestion
Furthermore, it will be possible to learn a separate dropout rate per layer, per neuron, or even per separate weight.
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+Despite this new approach, they have found that large values of $$\alpha$$ will imply large-variance gradients. 
+That's why they apply a constraint on $$\alpha$$ ($$\alpha \le 1$$ i.e.* $$p \le 0.5$$) during training.
+
+We will show in future post on sparse variational dropout that this constraint can be removed and induce sparse representation of a neural network.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:202]: Phrasing:
```suggestion
We will show in future post on sparse variational dropout that this constraint can be removed, leading to sparse representation of neural networks.
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+&nbsp;
+
+
+# Results
+
+Comparison of their method to standard binary dropout and two versions of Gaussian dropout.
+- Gaussian dropout A : pre-linear Gaussian dropout (based on a paper)
+- Gaussian dropout B : post-linear Gaussian dropout (based on another paper)
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:211]: It could be interesting to add references to the paper directly here, instead of having to look at the paper itself to find the references to these papers.

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 15]
```diff
@@ -0,0 +1,243 @@
...
+
+![](/collections/images/variational_dropout/img1.jpg)
+
+
+&nbsp;
+
+
+# Conclusion
+
+- They have obtain an efficient estimator with a low computational complexity and a low variance by injecting noise locally instead of globally.
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:229]: Conjugation:
```suggestion
- They obtained an efficient estimator with a low computational complexity and a low variance by injecting noise locally instead of globally.
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+&nbsp;
+
+
+## Local Reparameterization Trick
+
+What they explain in section 2.1 and 2.2 in the paper is that they can estimate the gradient of the log-likelyhood with Monte-Carlo based method. 
+It is also possible to do it to approximate the KL divergence. 
+But in practice the performance of stochastic gradient ascent crucially depends on the variance of the gradients. 
+They show that the variance of $$L_{\mathcal{D}}^{SGVB}(\phi)$$  can be dominated by the covariances for even moderately large $$M$$ (size of the minibatch).
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:86]: It would help to define the SGVB acronym before it is first used here.

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+&nbsp;
+
+
+
+
+
+
+
+
+
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:243]: Remove useless trailing blank lines:
```suggestion
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 18]
```diff
@@ -0,0 +1,243 @@
+---
+layout: review
+title: "Variational Dropout and the Local Reparameterization Trick"
+tags: Bayesian Inference, Distribution, Dropout,
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:4]: Format tags (because tags should not include space and are simply separated by white space):
```suggestion
tags: bayesian-inference distribution dropout
```

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 19]
```diff
@@ -0,0 +1,243 @@
+---
+layout: review
+title: "Variational Dropout and the Local Reparameterization Trick"
+tags: Bayesian Inference, Distribution, Dropout, 
+author: "Romain Deleat-besson"
+cite:
+    authors: "Kingma, D. P., Salimans, T., & Welling, M. "
+    title:   "Variational Dropout and the Local Reparameterization Trick"
+    venue:   "Advances in neural information processing systems (NeurIPS), 2015"
+pdf: "http://www.pachecoj.com/courses/csc665-1/papers/kingma15.pdf"
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:10]: I think it's a better idea to link to the paper hosted on NeurIPS' own website:
```suggestion
pdf: "https://proceedings.neurips.cc/paper/2015/file/bc7316929fe1545bf0b98d114ee3ecb8-Paper.pdf"
```

**Review by nathanpainchaud — CHANGES_REQUESTED** [review]
Great review of quite hard to grasp theoretical notions. Just a few minor edits I would recommend:

**Comment by RomainDeleat-besson** [comment]
Thank very much for your review Nathan !

File: collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md [hunk 20]
```diff
@@ -0,0 +1,242 @@
...
+
+&nbsp;
+
+
+## Local Reparameterization Trick
+
+What they explain in section 2.1 and 2.2 in the paper is that they can estimate the gradient of the log-likelyhood with Monte-Carlo based method. 
+It is also possible to do it to approximate the KL divergence. 
+But in practice the performance of stochastic gradient ascent crucially depends on the variance of the gradients. 
+They show that the variance of $$L_{\mathcal{D}}^{SGVB}(\phi)$$ (SGVB stands for Stochastic Gradient Variational Bayes define in their previous paper [here](https://arxiv.org/pdf/1312.6114.pdf?source=post_page---------------------------)) can be dominated by the covariances for even moderately large $$M$$ (size of the minibatch).
```
Comment (nathanpainchaud) [inline collections/_posts/2023-03-08-Variationnal_Dropout_and_the_local_reparameterization_trick.md:85]: Standarize URL linking to previous paper + formatting:
```suggestion
They show that the variance of $$L_{\mathcal{D}}^{SGVB}(\phi)$$ (where SGVB stands for _Stochastic Gradient Variational Bayes_, defined in their previous paper [here](https://arxiv.org/pdf/1312.6114.pdf)) can be dominated by the covariances for even moderately large $$M$$ (size of the minibatch).
```

**Review by nathanpainchaud — CHANGES_REQUESTED** [review]
Just a few follow-up comments regarding the formatting of the links + some left-over typos. But we're nearly good to go :wink:

**Comment by RomainDeleat-besson** [comment]
I've done the modifications you asked.
Thanks again for your time !

**Review by nathanpainchaud — APPROVED** [review]
